007_smoothed_Laplace/smoothed_laplace.ipynb

```diff
@@ -0,0 +1,598 @@
+    "#  See the License for the specific language governing permissions and\n",
+    "#  limitations under the License.\n",
+    "#\n",
+    "#  Authored by: Chao Zhang (DTU)\n",
+    "#  Reviewed by: Jakob Sauer Jørgensen (DTU)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Efficient sampling with Laplace priors\n",
+    "\n",
+    "### Chao Zhang (chaz@dtu.dk)\n",
+    "In this notebook, we explore methods for efficient sampling in solving inverse problems with Laplace priors. The Laplace distribution is important in situations where signals exhibit heavy tails or where outliers are more frequent than would be expected under a Gaussian distribution. It is also closely related to the LASSO (Least Absolute Shrinkage and Selection Operator) method in optimization that is used to induce sparsity in signals.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import numpy as np\n",
+    "import matplotlib.pyplot as plt\n",
+    "import cuqi\n",
+    "from cuqi.distribution import Gaussian, JointDistribution"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Set seed\n",
+    "np.random.seed(0)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "cuqi version:  1.0.0\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(\"cuqi version: \", cuqi.__version__)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 1. Introduction\n",
+    "\n",
+    "### Difficulty with Laplace priors\n",
+    "\n",
+    "The probability density function of a Laplace distribution is of form\n",
+    "$$p(x|\\mu, b)\\propto\\exp(-\\frac{|x-\\mu|}{b}),$$\n",
+    "where $\\mu$ is the mean and $b$ is the scale. The density function is not differentiable arround $\\mu$, and this is not desirable since it keeps us from using gradient-based samplers.\n",
+    "\n",
+    "### A remedy\n",
+    "\n",
+    "To overcome this issue, a remedy would be to \"smooth\" the Laplace distribution by introducing a smoothing parameter $\\epsilon>0$ and the density function becomes\n",
+    "$$p(x|\\mu, b)\\propto\\exp(-\\frac{\\sqrt{(x-\\mu)^2 + \\epsilon}}{b}).$$\n",
+    "\n",
+    "In doing so, we can make the density function differentiable at $\\mu$ and thus we can use gradient-based samplers.\n",
+    "\n",
+    "In CUQIpy, we can define such a smoothed Laplace distribution by subclassing `cuqi.distribution.Laplace` where we overwrite the `logpdf` and `gradient` methods. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class LaplaceSmoothed(cuqi.distribution.Laplace):\n",
```

Comment:
While in a way elegant to subclass to Laplace, I think probably better to make a subclass directly from distribution. Being subclassed from Laplace, it would inherit its `sample` method, and as such, would generate samples from the actual Laplace and not the smoothed version, which is not desired.

Reply:
Done